2/homework/laughter_prediction/rnn.py

```diff
@@ -0,0 +1,39 @@
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+
+
+class LibrosaFeaturesRnn(nn.Module):
+    def __init__(self, fbank_features, mfcc_features, hidden_size, batch_size=1):
+        super(LibrosaFeaturesRnn, self).__init__()
+
+        self.fbank_features = fbank_features
+        self.mfcc_features = mfcc_features
+        self.hidden_size = hidden_size
+        self.batch_size = batch_size
+
+        self.fbank_lstm = nn.LSTM(fbank_features, hidden_size)
+        self.mfcc_lstm  = nn.LSTM(mfcc_features, hidden_size)
+        self.linear_both = nn.Linear(hidden_size * 2, 2)
+        self.linear_mfcc = nn.Linear(hidden_size, 2)
+
+    def init_hidden_state(self, device):
+        return (torch.randn(1, self.batch_size, self.hidden_size).to(device),
+                torch.randn(1, self.batch_size, self.hidden_size).to(device))
+
+    def forward(self, x, device='cpu'):
+        x_fbank = x[:, :, :self.fbank_features].permute(1, 0, 2)
+        x_mfcc  = x[:, :, self.fbank_features:].permute(1, 0, 2)
+
+        out_fbank_lstm, _ = self.fbank_lstm(x_fbank, self.init_hidden_state(device))
```

Comment:
@maxkvant Why do you always initialize lstm hidden state with random value when you should pass the hidden from previous pass?

Reply:
I know 4 options for lstm initializartion:
1. zero state
2. random state
3. state from previous pass
4. noised state from previous pass

I have doubts about how to do approach 3 correctly for batches (Should I average it or not?). That's why I didn't consider options 3 and 4.

According to [post](https://r2rt.com/non-zero-initial-states-for-recurrent-neural-networks.html) random state better than zero state.